content/posts/a_confusing_build_failure.md

```diff
@@ -0,0 +1,64 @@
+---
+title: "A confusing build failure"
+date: 2019-11-07T01:00:09+02:00
+draft: true
+tags: ["testing", "continuous integration"]
+---
+
+{{< figure src="/images/posts/broken_builds/judge-dredd-under-arrest.jpg" title="" >}}
+
+Nowadays it has become quite a common practice to build and check software on machines other than local workstations, most of the time called CI servers(CI stands for [Continuous Integration](https://martinfowler.com/articles/continuousIntegration.html)). This can be done manually or automatically, usually triggered by a pull/merge request or a direct push on the main trunk(in git terminology most of the times named *master*).
+
+The point is that there's a single point where binaries are packaged and code is checked for quality and correctness, so we can merge changes safely and with confidence. 
+
+When a failure appears in a build it should be investigated why did it happen. Did a commit actually break functionality/introduce a bug? Was static code analysis checks too harsh and maybe they must be relaxed a little bit? One of the failures that left me most perplexed was a build that was failing on the CI server unit tests but passed locally! How did this happen? Investigating on this confusing build failure unearthed bigger problems than just a breaking unit
+test.
+
+## Mockito cannot mock final classes
+
+So why was the build failing? When the unit tests ran some of them would fail with an error stating that Mockito(more specifically [mockito-kotlin](https://github.com/nhaarman/mockito-kotlin)) - a library commonly used in Java and Android to provide mock test doubles and for behavior verification - cannot mock final classes.
```

Comment:
```suggestion
So why was the build failing? When the unit tests ran, some of them would fail with an error stating that Mockito (more specifically [mockito-kotlin](https://github.com/nhaarman/mockito-kotlin)) - a library commonly used in Java and Android to provide mock test doubles and behavior verification - cannot mock final classes.
```